test/MatrixFields/scalar_fieldmatrix.jl

```diff
@@ -0,0 +1,218 @@
+    end
+    struct Two_fields{T1, T2}
+        x::T1
+        y::T2
+    end
+    function test_broadcasted_get_field_type(
+        name,
+        ::Type{T},
+        expected_type,
+    ) where {T}
+        @test_all MatrixFields.broadcasted_get_field_type(T, name) ==
+                  expected_type
+    end
+    test_broadcasted_get_field_type(
+        @name(x),
+        Singleton{Singleton{Singleton{Singleton{Float64}}}},
+        Singleton{Singleton{Singleton{Float64}}},
+    )
+    test_broadcasted_get_field_type(
+        @name(x.x.x),
+        Singleton{Singleton{Singleton{Singleton{Float64}}}},
+        Singleton{Float64},
+    )
+    test_broadcasted_get_field_type(
+        @name(y.x),
+        Two_fields{
+            Two_fields{Float64, Float64},
+            Two_fields{Float64, Two_fields{Float64, Singleton{Float64}}},
+        },
+        Float64,
+    )
+    test_broadcasted_get_field_type(
+        @name(y.y.y),
+        Two_fields{
+            Two_fields{Float64, Float64},
+            Two_fields{Float64, Two_fields{Float64, Singleton{Float64}}},
+        },
+        Singleton{Float64},
+    )
+end
+
+@testset "fieldmatrix to scalar fieldmatrix unit tests" begin
+    FT = Float64
+    center_space, face_space = test_spaces(FT)
+    surface_space = Spaces.level(face_space, half)
+    seed!(1)
+    λ = 10
+    ᶜᶜmat1 = random_field(DiagonalMatrixRow{FT}, center_space) ./ λ .+ (I,)
+    ᶜᶠmat2 = random_field(BidiagonalMatrixRow{FT}, center_space) ./ λ
+    ᶠᶜmat2 = random_field(BidiagonalMatrixRow{FT}, face_space) ./ λ
+    ᶜᶜmat3 = random_field(TridiagonalMatrixRow{FT}, center_space) ./ λ .+ (I,)
+    ᶠᶠmat3 = random_field(TridiagonalMatrixRow{FT}, face_space) ./ λ .+ (I,)
+
+    e¹² = Geometry.Covariant12Vector(1, 2)
+    e³ = Geometry.Covariant3Vector(1)
+    e₃ = Geometry.Contravariant3Vector(1)
+
+    ρχ_unit = (; ρq_tot = 1, ρq_liq = 1, ρq_ice = 1, ρq_rai = 1, ρq_sno = 1)
+    ρaχ_unit =
+        (; ρaq_tot = 1, ρaq_liq = 1, ρaq_ice = 1, ρaq_rai = 1, ρaq_sno = 1)
+
+
+    ᶠᶜmat2_u₃_scalar = ᶠᶜmat2 .* (e³,)
+    ᶜᶠmat2_scalar_u₃ = ᶜᶠmat2 .* (e₃',)
+    ᶠᶠmat3_u₃_u₃ = ᶠᶠmat3 .* (e³ * e₃',)
+    ᶜᶠmat2_ρχ_u₃ = map(Base.Fix1(map, Base.Fix2(⊠, ρχ_unit ⊠ e₃')), ᶜᶠmat2)
+    A = dycore_prognostic_EDMF_FieldMatrix(;
+        ᶜᶜmat1,
+        ᶜᶠmat2,
+        ᶠᶜmat2,
+        ᶜᶜmat3,
+        ᶠᶠmat3,
+        e¹²,
+        e³,
+        e₃,
+        ρχ_unit,
+        ρaχ_unit,
+        ᶜᶠmat2_ρχ_u₃,
+        ᶠᶠmat3_u₃_u₃,
+        ᶜᶠmat2_scalar_u₃,
+        ᶠᶜmat2_u₃_scalar,
+    )
+
+    scalar_keys = MatrixFields.get_scalar_keys(A)
+    @test length(scalar_keys) > length(keys(A))
+    @test all(
+        f -> f isa MatrixFields.UniformScaling || eltype(eltype(f)) <: FT,
+        getindex.(Ref(A), scalar_keys),
+    )
+    @test (@allocated MatrixFields.get_scalar_keys(A)) == 0
+    @test_opt MatrixFields.get_scalar_keys(A)
+    foreach(scalar_keys) do key
+        @test_opt A[key]
+        @test (@allocated A[key]) == 0
+    end
```

Comment:
```suggestion
    @test_all length(MatrixFields.get_scalar_keys(A)) > length(keys(A))
    for key in MatrixFields.get_scalar_keys(A)
        @test_all A[key] isa MatrixFields.UniformScaling ||
            eltype(eltype(A[key])) == eltype(parent(A[key]))
    end
```
This could be a little less wordy if you use `@test_all` instead of writing separate tests for allocations and instabilities.